src/java/com/zimbra/oauth/handlers/impl/LinkedinOAuth2Handler.java

```diff
@@ -0,0 +1,231 @@
+        if (json.has(LinkedinMeConstants.FIRST_NAME.getValue()) && json.has(LinkedinMeConstants.LAST_NAME.getValue())
+                && !json.get(LinkedinMeConstants.FIRST_NAME.getValue()).asText().isEmpty()
+                && !json.get(LinkedinMeConstants.LAST_NAME.getValue()).asText().isEmpty()) {
+            return json.get(LinkedinMeConstants.FIRST_NAME.getValue()).asText() + "." + json.get(LinkedinMeConstants.LAST_NAME.getValue()).asText();
+        } else if (json.has(LinkedinMeConstants.ID.getValue()) && !json.get(LinkedinMeConstants.ID.getValue()).asText().isEmpty()) {
+            return json.get(LinkedinMeConstants.ID.getValue()).asText();
+        }
+
+        // if we couldn't retrieve the user first & last name, the response from
+        // downstream is missing data
+        // this could be the result of a misconfigured application id/secret
+        // (not enough scopes)
+        ZimbraLog.extensions.error("The user id could not be retrieved from the social service api.");
+        throw ServiceException.UNSUPPORTED();
+    }
+
+    @Override
+    public Boolean authenticate(OAuthInfo oauthInfo) throws ServiceException {
+        final Account account = oauthInfo.getAccount();
+        final String clientId = config.getString(
+            String.format(OAuth2ConfigConstants.LC_OAUTH_CLIENT_ID_TEMPLATE.getValue(), client), client,
+            account);
+        final String clientSecret = config.getString(
+            String.format(OAuth2ConfigConstants.LC_OAUTH_CLIENT_SECRET_TEMPLATE.getValue(), client),
+            client, account);
+        final String clientRedirectUri = config.getString(
+            String.format(OAuth2ConfigConstants.LC_OAUTH_CLIENT_REDIRECT_URI_TEMPLATE.getValue(), client),
+            client, account);
+        if (StringUtils.isEmpty(clientId) || StringUtils.isEmpty(clientSecret)
+            || StringUtils.isEmpty(clientRedirectUri)) {
+            throw ServiceException.FAILURE("Required config(id, secret and redirectUri) parameters are not provided.", null);
+        }
+        final String basicToken = OAuth2Utilities.encodeBasicHeader(clientId, clientSecret);
+        // set client specific properties
+        oauthInfo.setClientId(clientId);
+        oauthInfo.setClientSecret(clientSecret);
+        oauthInfo.setClientRedirectUri(clientRedirectUri);
+        oauthInfo.setTokenUrl(authenticateUri);
+        // request credentials from social service
+        final JsonNode credentials = getTokenRequest(oauthInfo, basicToken);
+        // ensure the response contains the necessary credentials
+        validateTokenResponse(credentials);
+        // determine account associated with credentials
+        final String username = getPrimaryEmail(credentials, account);
+        ZimbraLog.extensions.trace("Authentication performed for:" + username);
+
+        // get zimbra mailbox
+        final ZMailbox mailbox = getZimbraMailbox(oauthInfo.getZmAuthToken());
+
+        // store refreshToken
```

Comment:
`// store accessToken`

May not be unreasonable to note that `linkedin tokens are short-lived only`. But that could be left for the import implementation where a small change may be necessary in `syncDatasource` to not set a polling time.

Reply:
// store accessToken